### ISSUE TITLE
Add torch.compile to optimizer.step()

### ISSUE THREAD

**Issue by IvanKobzarev** [pytorch/torchtune#2516]
optimizer.step() is compilable for most of optimizers in moe.py
For cpu_offload case 17B model it helps to make 4 -> 10 tokens per second

```
--- a/torchtune/modules/moe/moe.py                                                                                                                                                                                                03:49:55 [25/39075]
+++ b/torchtune/modules/moe/moe.py
                             if isinstance(grad_norm, DTensor):
                                 grad_norm = grad_norm.full_tensor()
-                        self._optimizer.step()
-                        self._optimizer.zero_grad(set_to_none=True)
+
+                        def optimizer_fn():
+                                self._optimizer.step()
+                                self._optimizer.zero_grad(set_to_none=True)
+                        torch.compile(optimizer_fn)()
```

<img width="469" alt="Image" src="https://github.com/user-attachments/assets/c778826c-b47f-4112-85cf-bc33ac2d3a9f" />

@ebsmothers @shunting314

**Comment by felipemello1** [comment]
nice, it looks like we need to add more flexibility to our compile config, e.g.:

```
compile:
   model: True
   loss: True
   optimizer_step: True
```

**Comment by ionutmodo** [comment]
I would like to add a related issue to torch compile: when using `compile=True` and learning rate warmup, the code breaks because the scheduler expects to find the `__func__` attribute, but in `torchtune` the step function is compiled **before** creating the scheduler and the `__func__` attribute does not exist anymore.

The solution is to move the compilation of step function after we create the scheduler. See more here: https://github.com/pytorch/torchtune/issues/2263#issuecomment-3174478217

Note this happens with `full_finetune_distributed.py`, but might be the case for other recipes, too
